File: kotlin/src/main/kotlin/io/digitalstream/advanced/Enums.kt
/**
 * Created by sebby on 9/25/16.
 */
package io.digitalstream.advanced

object Enums {

  sealed class Option<T> {
    class None<T> : Option<T>()
    class Some<T>(val value: T) : Option<T>()
  }

  fun run(): Unit {
    val option = Option.Some("string")
  }
}

File: scala/src/main/scala/io/digitalstream/advanced/Enums.scala
/**
  * Created by sebby on 9/25/16.
  */
package io.digitalstream.advanced

object Enums {

  sealed trait Option[+A]

  case object None extends Option[Nothing]

  case class Some[+A](get: A) extends Option[A]

  def run(): Unit = {
    val option = Some("string")
  }
}
